easybuild/easyblocks/generic/cargo.py

```diff
@@ -0,0 +1,136 @@
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+EasyBuild support for installing Cargo packages (Rust lang package system)
+
+@author: Mikael Oehman (Chalmers University of Technology)
+"""
+
+import os
+
+import easybuild.tools.environment as env
+from easybuild.framework.easyconfig import CUSTOM
+from easybuild.framework.easyblock import EasyBlock
+from easybuild.tools.run import run_cmd
+from easybuild.tools.config import build_option
+from easybuild.tools.filetools import write_file, compute_checksum
+
+CRATESIO_SOURCE = "https://crates.io/api/v1/crates"
+
+
+class Cargo(EasyBlock):
+    """Support for installing Cargo packages (Rust)"""
+
+    @staticmethod
+    def extra_options(extra_vars=None):
+        """Define extra easyconfig parameters specific to Cargo"""
+        extra_vars = EasyBlock.extra_options(extra_vars)
+        extra_vars.update({
+            'enable_tests': [True, "Enable building of tests", CUSTOM],
+            'offline': [True, "Build offline", CUSTOM],
+            'lto': [False, "Build with link time optimization", CUSTOM],
+            'crates': [[], "List of (crate, version) tuples to use", CUSTOM],
+        })
+
+        return extra_vars
+
+    def __init__(self, *args, **kwargs):
+        """Constructor for Cargo easyblock."""
+        super(Cargo, self).__init__(*args, **kwargs)
+        env.setvar('CARGO_HOME', os.path.join(self.builddir, '.cargo'))
+        env.setvar('RUSTC', 'rustc')
+        env.setvar('RUSTDOC', 'rustdoc')
+        env.setvar('RUSTFMT', 'rustfmt')
+        optarch = build_option('optarch')
+        if not optarch:
+            optarch = 'native'
+        env.setvar('RUSTFLAGS', '-C target-cpu=%s' % optarch)
+        env.setvar('RUST_LOG', 'DEBUG')
+        env.setvar('RUST_BACKTRACE', '1')
+
+        # Populate sources from "crates" list of tuples
+        sources = self.cfg['sources']
+        for crate, version in self.cfg['crates']:
+            sources.append({
+                'download_filename': crate + '/' + version + '/download',
+                'filename': crate + '-' + version + '.tar.gz',
+                'source_urls': [CRATESIO_SOURCE],
+                'alt_location': 'crates.io',
+            })
+        self.cfg.update('sources', sources)
+
+    def configure_step(self):
+        pass
+
+    def extract_step(self):
+        """Populate all vendored deps with required .cargo-checksum.json"""
+        EasyBlock.extract_step(self)
+        for source in self.src:
+            dirname = source["name"].rsplit('.', maxsplit=2)[0]
```

Comment:
Crap.. turns out i can't rely on this. When i download sources from non-crates.io, this won't match, and it incorrectly creates a different directory. I need to know what the main directory is going to be.

Alternatively, i need to differentiate all the crates and control where they are extracted, to put all the vendored sources into a particular sub-directory.. which is going to be annoying to handle in the case where we are building a crates.io package itself (like ripgrep)